### DISCUSSION TITLE
Error when trying to connect my Node.JS app to the server

### DISCUSSION THREAD

**Discussion by remoteconn-7891** [community/community#75888] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Question

### Body

So I've asked this question twice on this forum with no solution for my problem. But I'm trying to send a connect to the server to the localhost. I've tried everything I can possibly think of to try and troubleshoot this issue.

I've replaced the 'localhost' with the IP address. Someone answered the previous question by suggesting to fix the code in package.JSON which didn't work either. So I'm gonna try this one last time on here, To make this simpler for everyone, I'm gonna copy/paste the code for all 4 files: package.JSON, index.js, auth.js and user.js as well as sending the screenshot of the error. These are all my node.js files.

package.JSON

`{
    "name": "Virtuallify",
    "version": "1.0.0",
    "description": "Marketplace & professional networking mobile app",
    "main": "index.js",
    "scripts": {
        "start": "node app.js",
        "test": "echo \"Error no test specified\" && exit  1"
    },
    "keywords": [],
    "author": "Corey James",
    "license": "ISC",
    "dependencies": {
        "bcryptjs": "^2.4.3",
        "express": "^4.18.2",
        "jsonwebtoken": "^9.0.2",
        "mongoose": "^8.0.0"
    },
    "devDependencies": {
        "nodemon": "^3.0.1"
    }
}
`

index.js

`var express = require("express");
var mongoose = require("mongoose");
var dotenv = require("dotenv"); // Import dotenv 
var authRouter = require("./routes/auth");
dotenv.config(); // Load environment variables // Load environment variables from .env file
var PORT = process.env.PORT || 3000;
var app = express();

app.use(json());
app.use(authRouter);
//DB just stands for MongoDB database, this String is the password for my database
const DB = 
"mongodb+srv://coreyjames689:remotify87@cluster0.jybj21u.mongodb.net/Cluster0?retryWrites=true&w=majority";

connect(DB)
  .then(() => {
    console.log("Connection Successful");
  })
  // catch(e) here is used for error handling by catching coding errors in Express
  .catch((e) => {
    console.log(e);
  });

  // My server is listening on PORT 3000
app.listen(3000, () => {
  console.log('Server is running on port ${(3000)}');
});`

user.js

`const mongoose = require("mongoose");

const userSchema = mongoose.Schema({
    name: {
        required: true,
        type: String,
        trim: true,
    },
    email: {
        required: true,
        type: String,
        trim: true,
        unique: true,
        validate: {
            validator: (value) => {
                // Regular expression for a simple email validation
                const re = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
                return re.test(value.toLowerCase()); // Convert to lowercase to ensure case-insensitive validation
            },
            message: "Invalid email format",
        },
    },
    password: {
        required: true,
        type: String,
        validate: {
            validator: (value) => {
                return value.length >= 8; // Minimum password length of 8 characters
            },
            message: "Password must be at least 8 characters long",
        },
    },
    role: {
        required: true,
        type: String,
        enum: ["job_seeker", "company"], // Enum to define possible roles
    },
});

const User = mongoose.model("User", userSchema);

module.exports = User;
`
auth.js

`const express = require("express");
const cors = require("cors");
//Using bodyParser as a Middleware function
const bodyParser = require('body-parser');
const bcryptjs = require("bcryptjs");
const authRouter = express.Router();

//Create an Account
authRouter.post("/ap/signup", async (req, res) => {
try {
    const { name, email, password, role } = req.body;
 
    const existingUser = await user.findOne({ email });
    if (existingUser) {
        return res
          .status(400)
          .json({ msg: "User with this email already exists." });
    }

    const hashedPassword = await bcryptjs.hash(password, 8);

    //User login credentials include user, email address, password. I also
    //included a role field to differentiate between remote job seeker and
    //company
  let user = new user({
    email,
    password: hashedPassword,
    name,
    role,
  });
  user = await user.save();
  res.json(user);
} catch (e) {
  res.status(500).json({ error: e.message });
}
});

// Example login route
authRouter.post("/login", async (req, res) => {
    const { email, password } = req.body;

    try {
        const user = await User.findOne({ email, password });
        //This 401 error code just insinuates that it failed to
        //authenticate the user's login credentials
        if (!user) {
            return res.status(401).json({ message: "Invalid credentials" });
        }

        // Check the user's role and redirect accordingly
        if (user.role === "job_seeker") {
            // Redirect to job seeker dashboard
            res.redirect("/job_seeker/dashboard");
        } else if (user.role === "company") {
            // Redirect to company dashboard
            res.redirect("/company/dashboard");
        } else {
            res.status(401).json({ message: "Invalid role" });
        }
    } catch (error) {
        console.error(error);
        res.status(500).json({ message: "Internal server error" });
    }
});

module.exports = authRouter;`

Please ignore the 'var' in the index.js, I changed that on purpose. Please let me know if you see errors in the code for any of these node.JS files. I'll send the screenshot on another post

**Comment by remoteconn-7891** [comment]
<img width="1440" alt="Screenshot 2023-11-17 at 11 42 32 AM" src="https://github.com/community/community/assets/142646120/38f689c8-668c-451e-9802-842c329e8786">

**Comment by ghost** [comment]
Hello @remoteconn-7891 ,
I have read your question and could not find any clear issue. 

However I can suggest to include better error description in this question. The error you provided does not include any information/detail for troubleshooting. 

If you have not checked CORS policy I suggest to check if its cousing  your issue. There can be headers problem (origin of request, content type and so on). 

Also you can try to send request to your server using platforms like Postman or Insomnia (if its headers issue then you should be able to get correct response from server using postman or at least better error description. it will be super helpfull if you get error code or anything related).

try to hardcode values of env variables for during debug.

try to log server errors in a file. create new file for logs and log every details of your server file (for example log "Get Request:" and request details where you listen for request). this will help you to have an idea where the error is happening.

Hope this helps.

**Comment by remoteconn-7891** [comment]
Okay, thank you. I decided to go through another online node.js Login tutorial with a boiler plate code. I'll have to customize both the email and password conditions

**Comment by remoteconn-7891** [comment]
By the way, can you possible send me some code snippets of email conditions, i.e. has to be jondoe@gmail.com for example and also maybe hashed password condition and minimum password length code for security? Or should I use ChatGPT?

**Reply by ghost** [reply]
do you mean regex to validate email and password?

**Reply by remoteconn-7891** [reply]
Don't know about regex. But yes, I'm asking about authenticating email and password requirements. So for example, a string of random characters for the email validation and hashed password conditions for security purposes

**Reply by remoteconn-7891** [reply]
Yes, with regex like this: var emailRegex = /^[-!#$%&'*+\/0-9=?A-Z^_a-z{|}~](\.?[-!#$%&'*+\/0-9=?A-Z^_a-z`{|}~])*@[a-zA-Z0-9](-*\.?[a-zA-Z0-9])*\.[a-zA-Z](-?[a-zA-Z0-9])+$/;

**Reply by ghost** [reply]
@remoteconn-7891 

so for gmail validation you can use this regex: `/^[\w-\.]+@([\w-]+\.)+[\w-]{2,4}$/`
for password validation it is little bit different since different platforms has diferent password requirements. so for example if you want user password to be at least 8 char + at least one upper case latter + at least one lower case latter + at least one number + at least one special symbol this is regex for it: `/^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@$!%*?&])[A-Za-z\d@$!%*?&]{8,}$/` (super safe :smile: )

you can find more variations of password regex [there](https://stackoverflow.com/questions/19605150/regex-for-password-must-contain-at-least-eight-characters-at-least-one-number-a) (on stackoverflow)

Also it is very hard to understand regex and if it works well or not. so if needed you can test your regex on that platform: [RegExer](https://regexr.com/3e48o)

**Comment by remoteconn-7891** [comment]
I'll try and send you the boilerplate codes tonight when I finish it

**Comment by remoteconn-7891** [comment]
So I finished the boilerplate codes for the login screen. I will copy/paste all files here.

index.js:

`const bodyParser = require("body-parser")
var express = require ("express")
const passport = require ("passport")
    mongoose = require ("mongoose"),
    passport = require ("passport"),
    bodyParser = require ("body-parser"),
    LocalStrategy = require ("passport-local"),
    passportLocalMongoose =
        require ("passport-local-mongoose")
const User = require ("./model/User");
const { default: mongoose } = require("mongoose");
var app = express();

mongoose.connect ("mongodb+srv://coreyjames689:HhLEuC0vhIeZIExE@cluster0.jybj21u.mongodb.net/?retryWrites=true&w=majority")

app.set ("view engine", "ejs");
app.use(bodyParser.urlencoded ({ extended: true }));
app.use(require ("express-session")) ({
    secret: "My name is Don",
    resave: false,
    saveUninitialized: false
});

app.use(passport.initialize());
app.use(passport.session());

passport.use(new LocalStrategy (User.authenticate()));
passport.serializeUser (User.serializeUser());
passport.deserializeUser (User.deserializeUser());

//=======
// Routes
//=======

// Home screen
app.get("/", function (require, res) {
    res.render("home");
});

// Secret screen
app.get("/secret", isLoggedIn, function (req, res) {
    res.render("secret");
});

// Create an account
app.get("/signup", async(req, res) => {
    const user = await User.create({
        email: req.body.email,
        password: req.body.password
    });

    return res.status(200).json(user);
});

// Sign-In Screen
app.get("/sign-in", function(req, res) {
    res.render("sign-in");
}); 

// User sign-in function
app.post("/sign-in", async function(req, res) {
    try {
        // validate existing user
        const user = await User.find0ne({ email: req.body.email });
        if (user) {
            // verify matching passwords
            const result = req.body.password === user.password;
            if (result) {
                res.render("secret");
            } else {
              res.status(400).json({ error: "passwords don't match" }); 
            }
        } else {
            res.status(400).json({ error: "User with this account doesn't exist" });
        }
    } catch (error) {
        res.status(400).json({ error });
    }
});

// User Sign-Out function
app.get("/signout", function (req, res) {
    re1.signout(function(err) {
        if (err) { return next(err); }
        res.redirect('/');
    });
});


function isLoggedIn(req, res, next) {
    if (req.isAuthenticated()) return next();
    res.redirect("/sign-in");
}
// Server is listening on PORT 3000
var port = process.env.PORT || 3000;
app.listen(port, function () {
    console.log("Server is listening on Port 3000");
});`

user.js:

`const mongoose = require('mongoose')
const Schema = mongoose.Schema
const passportLocalMongoose = require('passport-local-mongoose');
var User = new Schema({
    email: {
        type: String
    },
    password: {
        type: String
    }
})

User.plugin(passportLocalMongoose);

module.exports = mongoose.model('User', User)`

index.js is the entry point node.js file and user.js inside model folder. Here are the ejs files for HTML inside view folder. 

home.ejs:

`<h1>This is the Home Page</h1>

<li><a href="/signup">Create an account!</a></li>
<li><a href="/sign-in">Sign In</a></li>
<li><a href="/signout">Sign Out</a></li>`

login.ejs:

`<h1>This is the Home Page</h1>

<li><a href="/signup">Create an account!</a></li>
<li><a href="/sign-in">Sign In</a></li>
<li><a href="/signout">Sign Out</a></li>`

signup.ejs:

`<h1> Register an account</h1>

<form action="/signup" method="POST">
    <input type="text" name="email"
        placeholder="email">
    <input type="password" name="password"
        placeholder="=password">
        <button>submit</button>
</form>`

secret.ejs:

`<h1> Register an account</h1>

<form action="/signup" method="POST">
    <input type="text" name="email"
        placeholder="email">
    <input type="password" name="password"
        placeholder="=password">
        <button>submit</button>
</form>`

secret.ejs:

`<h1>This is a secret page</h1>

<img src=
"https://media.geeksforgeeks.org/wp-content/cdn-uploads/20210420155809/gfg-new-logo.png"

`
I will send you a screenshot of the error when I type the URL 'http://localhost:3000/'. I think I coded it correctly, it's slightly modified as I replaced username with email. PS I'm also using PassPort

**Comment by remoteconn-7891** [comment]
<img width="1440" alt="Screenshot 2023-11-18 at 7 26 15 PM" src="https://github.com/community/community/assets/142646120/d860a720-e4c1-4d92-a836-7b0e192b433c">

**Comment by remoteconn-7891** [comment]
Sorry I forgot to paste the package.JSON code here:

`{
  "name": "virtualify", // project name
  "version": "1.0.0",
  "main": "index.js",
  "scripts": {
    "start": "node index.js"
  },

  "keywords": [],
  "author": "Corey",
  "license": "MIT",
  "description": "maretplace and professional networking mobile app",
  "dependencies": {
    "body-parser": "^1.20.2",
    "ejs": "^3.1.9",
    "express": "^4.18.2",
    "express-session": "^1.17.3",
    "mongoose": "^8.0.1",
    "passport": "^0.6.0",
    "passport-local": "^1.0.0",
    "passport-local-mongoose": "^8.0.0"
  }
}`

**Reply by Senocart3** [reply]
Hinterlossen

**Reply by remoteconn-7891** [reply]
What?

**Reply by Senocart3** [reply]
> Sorry I forgot to paste the package.JSON code here:
> 
> 
> 
> `{
> 
>   "name": "virtualify", // project name
> 
>   "version": "1.0.0",
> 
>   "main": "index.js",
> 
>   "scripts": {
> 
>     "start": "node index.js"
> 
>   },
> 
> 
> 
>   "keywords": [],
> 
>   "author": "Corey",
> 
>   "license": "MIT",
> 
>   "description": "maretplace and professional networking mobile app",
> 
>   "dependencies": {
> 
>     "body-parser": "^1.20.2",
> 
>     "ejs": "^3.1.9",
> 
>     "express": "^4.18.2",
> 
>     "express-session": "^1.17.3",
> 
>     "mongoose": "^8.0.1",
> 
>     "passport": "^0.6.0",
> 
>     "passport-local": "^1.0.0",
> 
>     "passport-local-mongoose": "^8.0.0"
> 
>   }
> 
> }`

Apache.org/licens/2.0 copyright [20231126] [Artur Vogiel] {Erlang} #339966#

**Comment by remoteconn-7891** [comment]
That's fine. Thank you, I will try that. Please keep in mind that I tweaked a few things in index.js. I'm just tired of relying on ChatGPT

**Comment by remoteconn-7891** [comment]
It didn't work. Here is an updated package.JSON 

` "name": "virtualify",
  "version": "1.0.0",
  "main": "index.js",
  "scripts": {
"start": "NODE_PATH=${PWD} node index.js"
  },
  "keywords": [],
  "author": "Corey",
  "license": "MIT",
  "description": "maretplace and professional networking mobile app",
  "dependencies": {
    "bcrypt": "^5.1.1",
    "body-parser": "^1.20.2",
    "ejs": "^3.1.9",
    "express": "^4.18.2",
    "express-session": "^1.17.3",
    "jsonwebtoken": "^9.0.2",
    "mongoose": "^8.0.1",
    "nodemon": "^3.0.1",
    "passport": "^0.6.0",
    "passport-local": "^1.0.0",
    "passport-local-mongoose": "^8.0.0",
    "react-native": "^0.72.7"
  }
}`
